core/src/main/scala/org/http4s/multipart/Part.scala

```diff
@@ -29,10 +34,41 @@ object Part {
   def empty[F[_]]: Part[F] =
     Part(Headers.empty, EmptyBody)
 
-  def formData[F[_]: Sync](name: String, value: String, headers: Header*): Part[F] =
+  def formData[F[_]: Sync](name: String, value: String, headers: Header*): Part[F] = {
+    val body = Stream.emit(value).through(utf8Encode)
+    val bodyLength = body.as(1).compile.fold(0L)(_ + _)
     Part(
       Headers(`Content-Disposition`("form-data", Map("name" -> name)) :: headers.toList),
-      Stream.emit(value).through(utf8Encode))
+      Stream.emit(value).through(utf8Encode),
+      Some(bodyLength))
+  }
+
+  def fileDataKnownContentLength[F[_]: Sync: ContextShift](
+      name: String,
+      path: Path,
+      blockingEC: ExecutionContext,
+      headers: Header*): F[Part[F]] =
+    fileDataKnownContentLength(
+      name,
+      path.toFile.getName,
+      readAll[F](path, blockingEC, ChunkSize),
```

Comment:
How about `F.delay(Files.size(path))` and flatMap that shit?  I guess there's a slight risk the file size changes between being sized and being read.

Reply:
Sounds pretty reasonable, though we can't really do that when all we have is an `entityBody`, do we?

Reply:
I think I wouldn't implement it in terms of the other.  I'm still a little scared of reading a growing file, where the size can become a lie.  And the javadoc is frightening:

> Returns the size of a file (in bytes). The size may differ from the actual size on the file system due to compression, support for sparse files, or other reasons. The size of files that are not regular files is implementation specific and therefore unspecified.

We could `.take(n)` bytes from the stream we read to guarantee we don't go over, but this would silently truncate things.  Maybe we have one like I'm proposing that's optimal, but with appropriate disclaimers, and then a buffering one like what you already wrote?

Reply:
Alright, thanks for the advice, will do :)